src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -572,37 +631,50 @@ GenerateSizeQueryOnMultiplePlacements(List *shardIntervalList, char *sizeQuery)
 
 
 /*
- * GenerateAllShardNameAndSizeQueryForNode generates a query that returns all
- * shard_name, shard_size pairs for the given node.
+ * GenerateAllShardStatisticsQueryForNode generates a query that returns:
+ * - all shard_name, shard_size pairs for the given node (if useShardMinMaxQuery is false)
+ * - all shard_id, shard_minvalue, shard_maxvalue, shard_size quartuples (if true)
  */
 static char *
-GenerateAllShardNameAndSizeQueryForNode(WorkerNode *workerNode)
+GenerateAllShardStatisticsQueryForNode(WorkerNode *workerNode, List *citusTableIds, bool
+									   useShardMinMaxQuery)
 {
-	List *allCitusTableIds = AllCitusTableIds();
-
-	StringInfo allShardNameAndSizeQuery = makeStringInfo();
+	StringInfo allShardStatisticsQuery = makeStringInfo();
 
 	Oid relationId = InvalidOid;
-	foreach_oid(relationId, allCitusTableIds)
+	foreach_oid(relationId, citusTableIds)
```

Comment:
we should have probably taken accessShare lock here and probably some other places: https://github.com/citusdata/citus/issues/5116